### PR TITLE
2019 10 24 | Signals /  Model Validations / SettingMiddleware / MultiSites Settings

### PR DESCRIPTION
**Pull request by iamhaiderkhan**
- Write  Post_save receiver for sending welcome email to new users
 - Model-level fields validations
 - Write Custom Setting Middleware for set multi-sites email settings on runtime.
    e.g : 
    edx.localhost 
    edly.localhost

### CODE REVIEW

File: user_profile/models.py [hunk 1]
```diff
@@ -15,4 +26,40 @@ class UserProfile(models.Model):
...
+                [instance.email],
+                fail_silently=False
+            )
+        except SMTPException:
+            pass
+        print(settings.EMAIL_HOST_USER)
+        print("Email Successfully Sent to:", instance.email)
+
+
+class Settings(models.Model):
```
Comment (ziafazal) [inline user_profile/models.py]: `Settings` is not a good name for model. Invitation for conflicts.

File: user_profile/models.py [hunk 2]
```diff
@@ -15,4 +26,40 @@ class UserProfile(models.Model):
...
+            send_mail(
+                'New Account',
+                'Welcome to Python training Application.',
+                settings.EMAIL_HOST_USER,
+                [instance.email],
+                fail_silently=False
+            )
+        except SMTPException:
+            pass
+        print(settings.EMAIL_HOST_USER)
```
Comment (ziafazal) [inline user_profile/models.py]: we should log things here instead of printing them.

File: user_profile/migrations/0006_auto_20191025_1315.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
+# Generated by Django 2.2.6 on 2019-10-25 13:15
+
+from django.db import migrations, models
```
Comment (ziafazal) [inline user_profile/migrations/0006_auto_20191025_1315.py]: merge migrations

File: user_profile/managers.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+        return self.count()
+
+
+class SettingsManager(Manager):
+
+    def  get_site_id(self, site_domain):
+        current_domain = Site.objects.get_by_natural_key(site_domain)
+        return current_domain.id
+
+    def get_current_settings(self, current_site_id):
```
Comment (ziafazal) [inline user_profile/managers.py]: both `get_site_id ` and  `get_current_settings` can be merged into a single method  something like 
https://docs.djangoproject.com/en/2.2/ref/contrib/sites/#django.contrib.sites.shortcuts.get_current_site
```
def get_settings_by_site(request)
current_site = get_current_site(request)
return self.get(site=current_site)
```

**Review by ziafazal — CHANGES_REQUESTED** [review]
@iamhaiderkhan reviewed

File: user_profile/models.py (hunk 1 above)
Comment (iamhaiderkhan) [reply user_profile/models.py]: Changed.

File: user_profile/models.py (hunk 2 above)
Comment (iamhaiderkhan) [reply user_profile/models.py]: Done.

File: user_profile/managers.py (hunk 4 above)
Comment (iamhaiderkhan) [reply user_profile/managers.py]: done

**Review by ziafazal — APPROVED** [review]
👍

**Pull request merged**